### PR TITLE
HTTP: add handlers to allow object creation and modification

### PR DESCRIPTION
**Pull request by mildred**
This is for issue #611

### CODE REVIEW

File: core/corehttp/gateway.go [hunk 1]
```diff
@@ -6,11 +6,13 @@ import (
...
-	if err != nil {
-		return err
+func GatewayOption(readOnly bool) ServeOption {
+	return func(n *core.IpfsNode, mux *http.ServeMux) error {
+		gateway, err := newGatewayHandler(n, readOnly)
+		if err != nil {
+			return err
+		}
+		mux.Handle("/ipfs/", gateway)
+		return nil
```
Comment (jbenet) [inline core/corehttp/gateway.go]: :+1: on this approach to options

File: cmd/ipfs/daemon.go [hunk 2]
```diff
@@ -36,6 +37,7 @@ the daemon.
 	Options: []cmds.Option{
 		cmds.BoolOption(initOptionKwd, "Initialize IPFS with default settings if not already initialized"),
 		cmds.BoolOption(mountKwd, "Mounts IPFS to the filesystem"),
+		cmds.BoolOption(readOnlyKwd, "Make the HTTP gateway read only (disable POST, PUT and DELETE)"),
```
Comment (jbenet) [inline cmd/ipfs/daemon.go]: let's flip this-- i think the default should be read only and writing should need enabling (at least for the time being). maybe:

```
-w --writable "Enable writing objects (with POST, PUT, and DELETE)"
```

File: core/corehttp/gateway_handler.go [hunk 3]
```diff
@@ -163,23 +194,170 @@ func (i *gatewayHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+
+	http.Redirect(w, r, "/ipfs/"+mh.Multihash(k).B58String(), http.StatusCreated)
+}
+
+func (i *gatewayHandler) putEmptyDirHandler(w http.ResponseWriter, r *http.Request, path string) {
+	newnode := NewDagEmptyDir()
+
+	key, err := i.node.DAG.Add(newnode)
+	if err != nil {
+		webError(w, "Could not add recursively new node", err, http.StatusInternalServerError)
```
Comment (jbenet) [inline core/corehttp/gateway_handler.go]: nitpick: `"Could not recursively add new node"`

File: core/corehttp/gateway_handler.go [hunk 4]
```diff
@@ -163,23 +194,170 @@ func (i *gatewayHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+func (i *gatewayHandler) putHandler(w http.ResponseWriter, r *http.Request, path string) {
+	var err error
+	if path == "/QmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn/" {
+		i.putEmptyDirHandler(w, r, path)
+		return
+	}
+
+	var newnode *dag.Node
+	if path[len(path)-1] == '/' {
+		newnode = NewDagEmptyDir()
```
Comment (jbenet) [inline core/corehttp/gateway_handler.go]: should there be an early return here? is it easy for this to get mishandled below if people modify the code?

File: core/corehttp/gateway_handler.go [hunk 5]
```diff
@@ -163,23 +194,170 @@ func (i *gatewayHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+	key, err := i.node.DAG.Add(newnode)
+	if err != nil {
+		webError(w, "Could not add recursively new node", err, http.StatusInternalServerError)
+		return
+	}
+
+	http.Redirect(w, r, "/ipfs/"+key.String()+"/", http.StatusCreated)
+}
+
+func (i *gatewayHandler) putHandler(w http.ResponseWriter, r *http.Request, path string) {
```
Comment (jbenet) [inline core/corehttp/gateway_handler.go]: this function is pretty long and there's lots of edge-case handling. Let's add some comments like so:

```go
// resolving path components into merkledag nodes
// if no links are found, create empty directories -- which will be linked and populated below.
path_nodes, err := i.node.Resolver.ResolveLinks(rootnd, components[:len(components)-1])
if _, ok := err.(p.ErrNoLink); ok {
  // Create empty directories, links will be made further down the code
  for len(path_nodes) < len(components) {
    path_nodes = append(path_nodes, NewDagEmptyDir())
  }
} else if err != nil {
  webError(w, "Could not resolve parent object", err, http.StatusBadRequest)
  return
}
```

File: merkledag/node.go [hunk 6]
```diff
@@ -134,6 +134,16 @@ func (n *Node) Copy() *Node {
...
 }
 
+// UpdateNodeLink return a copy of the node with the link name set to point to
+// that. If a link of the same name existed, it is removed.
+func (n *Node) UpdateNodeLink(name string, that *Node) (*Node, error) {
+	newnode := n.Copy()
+	err := newnode.RemoveNodeLink(name)
+	err = nil // ignore error
+	err = newnode.AddNodeLink(name, that)
+	return newnode, err
```
Comment (jbenet) [inline merkledag/node.go]: hmmm not sure if we want ipns to be mutable. I dont know this part of the codebase very well. I'll let @whyrusleeping shepherd the IPNS parts here

File: path/path.go [hunk 7]
```diff
@@ -13,16 +13,25 @@ import (
...
 	DAG merkledag.DAGService
 }
 
-// ResolvePath fetches the node for given path. It uses the first
-// path component as a hash (key) of the first node, then resolves
-// all other components walking the links, with ResolveLinks.
-func (s *Resolver) ResolvePath(fpath string) (*merkledag.Node, error) {
+// SplitPath clean up and split fpath. It extracts the first component (which
+// must be a Multihash) and return it separately.
+func SplitPath(fpath string) (mh.Multihash, []string, error) {
```
Comment (jbenet) [inline path/path.go]: Maybe `SplitAbsPath` will make the semantics clearer to readers

File: path/path.go [hunk 8]
```diff
@@ -39,13 +48,36 @@ func (s *Resolver) ResolvePath(fpath string) (*merkledag.Node, error) {
...
+
+// ResolvePath fetches the node for given path. It returns the last item
+// returned by ResolvePathComponents.
+func (s *Resolver) ResolvePath(fpath string) (*merkledag.Node, error) {
+	nodes, err := s.ResolvePathComponents(fpath)
+	if nodes == nil {
+		return nil, err
+	} else {
+		return nodes[len(nodes)-1], err
+	}
```
Comment (jbenet) [inline path/path.go]: always check for `err != nil` because the other thing may return `err` even with partially valid output (and in most cases we'll want to suppress valid outputs when there are errors). I get the intent to check for nil on the slice... maybe:
```
if err != nil || nodes ==  nil {
    return nil, err
}
return nodes[len(nodes)-1], nil
```

**Comment by jbenet** [comment]
@mildred thanks for the PR! a few comments above:
- [ ] switch mode option `readable --> writable`
- [ ] more comments on the putHandler
- [ ] `SplitAbsPath`
- [ ] small nit picks

And, @mappum and @whyrusleeping can you CR the gateway + ipns parts? (i dont have all that state as fresh in my mind so you'll be able to decide correctness better)

File: merkledag/node.go (hunk 6 above)
Comment (mildred) [inline merkledag/node.go]: That's why the Node is not mutated by the function. It is copied on a new instance which is modified. In fact, that was already done like that in another part of the code (`update()` function in ` fuse/ipns/ipns_unix.go`).

**Comment by mildred** [comment]
I am going to update the code to fix the small issues you reported. Thank you for taking time to review my code.

File: core/corehttp/gateway_handler.go (hunk 4 above)
Comment (mildred) [inline core/corehttp/gateway_handler.go]: It's not really possible to make this condition an early return. The new node still need to be saved with its parent directories. And there is semantically no difference ahead in code. We have a new node, and the rest of the function is to update the parent nodes.

**Comment by mildred** [comment]
I rebased the branch fixing the things you noticed.

File: merkledag/node.go (hunk 6 above)
Comment (jbenet) [inline merkledag/node.go]: Ahhh yes, i had missed the `n.Copy()`

**Comment by jbenet** [comment]
Modifications LGTM-- @mappum @whyrusleeping your parts?

**Comment by mappum** [comment]
LGTM, this will be a very cool API for building DAG objects. But there be some conflicts to resolve when rebasing, sorry.

BTW I'll probably add an API that uses this in my JS library.

File: core/corehttp/gateway.go [hunk 9]
```diff
@@ -6,11 +6,13 @@ import (
...
-	gateway, err := newGatewayHandler(n)
-	if err != nil {
-		return err
+func GatewayOption(writable bool) ServeOption {
+	return func(n *core.IpfsNode, mux *http.ServeMux) error {
+		gateway, err := newGatewayHandler(n, writable)
+		if err != nil {
+			return err
+		}
+		mux.Handle("/ipfs/", gateway)
```
Comment (briantigerchow) [inline core/corehttp/gateway.go]: There's some complexity emerging in `ServeHTTP`. There may be a simpler way to handle multiple HTTP methods on the same path. If we switch over to the gorilla mux, we get the following...

```Go
state := &gatewayState{node: n, writable: writable, dir: load()}
ipfsPath := r.PathPrefix("/ipfs/").Subrouter()
ipfsPath.Methods("GET").HandlerFunc(getHandler(state))
ipfsPath.Methods("PUT").HandlerFunc(putHandler(state))
ipfsPath.Methods("POST").HandlerFunc(postHandler(state))
ipfsPath.Methods("DELETE").HandlerFunc(deleteHandler(state))
```

Thoughts?

Comment (jbenet) [inline core/corehttp/gateway.go]: I'm in favor of moving to `gorilla.Mux`

File: core/corehttp/gateway_handler.go [hunk 10]
```diff
@@ -14,7 +15,9 @@ import (
 	"github.com/jbenet/go-ipfs/importer"
 	chunk "github.com/jbenet/go-ipfs/importer/chunk"
 	dag "github.com/jbenet/go-ipfs/merkledag"
+	p "github.com/jbenet/go-ipfs/path"
```
Comment (jbenet) [inline core/corehttp/gateway_handler.go]: longer import name -- `p` is too unrecognizable for comfort

**Comment by jbenet** [comment]
Hey @mildred the major thing on my end is tests -- I want to get IPFS to a git-like reliability level. Being careful to test everything is what will get us there. -- i worry that including this major piece of functionality without tests might regress quickly (there's lots of different cases in `putHandler` and `deleteHandler` to take into account. Are you familiar with the sharness tests? @mappum has started writing some with `curl` for the gateway in a branch. this might be good here.  If you need this in sooner, i'd be fine merging and adding those in a separate PR, but ideally would go in here. Thoughts?

File: path/path.go [hunk 11]
```diff
@@ -55,19 +87,22 @@ func (s *Resolver) ResolvePath(fpath string) (*merkledag.Node, error) {
...
 
 // ResolveLinks iteratively resolves names by walking the link hierarchy.
 // Every node is fetched from the DAGService, resolving the next name.
-// Returns the last node found.
+// Returns the list of nodes forming the path, starting with ndd. This list is
+// guaranteed never to be empty.
 //
 // ResolveLinks(nd, []string{"foo", "bar", "baz"})
 // would retrieve "baz" in ("bar" in ("foo" in nd.Links).Links).Links
 func (s *Resolver) ResolveLinks(ndd *merkledag.Node, names []string) (
```
Comment (whyrusleeping) [inline path/path.go]: Why was this change necessary?

Comment (mildred) [inline path/path.go]: We need this because when we PUT to /ipfs/HASH/dir/dir/file, the hash of file changes. This makes necessary to update the directories in succession up to the root hash. I felt it was better not to parse the path two times (once for getting the leaf file hash, one other time to rewrite the directories). Instead the path parsing is done once using the ResolveLinks function

**Comment by mildred** [comment]
Thanks jbenet for your reply. Indeed, I think tests are a very good addition. Unfortunately, I'm not too skilled in them, but I'll have to learn.

Do you think tests that are run from `go test` are enough or should we really test from the outside using curl ? I'd go for the former, and I'll update my pull request. I'm currently learning to do tests in my Ed25519 PR (#677) and they are really useful and easy to do, once you know what to test.

**Comment by jbenet** [comment]
@mildred the more tests the better. I think for something like this (an http api), we'll always want end-to-end tests (testing it as a black box). We can do those in Go, using the `http` package and doing Gets and Puts.  if we're careful about the tests written, go tests can be fine. etc. doing them in sharness makes it obvious that something really works outside of go + the same process, and ensures that the formats are all correct (sharness makes it very hard to trick yourself).

The sharness tests are easy to pick up, too. take a look at https://github.com/jbenet/go-ipfs/blob/master/test/sharness/t0080-repo.sh

File: path/path.go (hunk 11 above)
Comment (jbenet) [inline path/path.go]: @whyrusleeping if you want the signature of `ResolveLinks` not to change, we can move this functionality to a differently-named function.

**Comment by mildred** [comment]
I updated the pull request. Removed complexity from the ServeHttp function, implemented /ipns GET requests (as this code was lying around me unused and was part of a simplification refactoring).

I did not switch to gorilla Mux, there seem to be different implementations around (one at github.com the other at code.google.com). And I think this might involve `gateway.go#GatewayOption` or `gateway_handler.go#newGatewayHandler` and switching away from the current Mux.

I still have to make some tests.

**Comment by jbenet** [comment]
@mildred i think @mappum rebased your branch for you (we merged more stuff into this area). what branch @mappum ?

**Comment by mappum** [comment]
I rebased here: https://github.com/jbenet/go-ipfs/commits/http-rest

The tests might need to go back to using curl so we can test PUT and DELETE, but otherwise it should be RFM.

**Comment by jbenet** [comment]
> The tests might need to go back to using curl 

is curl available everywhere? (we could install it with make)

**Comment by mildred** [comment]
I also updated this branch on my side: see https://github.com/mildred/go-ipfs/commits/http-rest-wip. Rebasing and adding sharness tests using curl and shell scripting (see https://github.com/mildred/go-ipfs/commit/3cb8dbd9d107b0f9dd2ce5b8e55a7e062bc20cd0). But I seem to get in conflict with pull request #674. So I'll take a little time to get it in working order.

**Comment by mildred** [comment]
I rebased my pull request. I fixed up the code and added sharness tests

**Comment by jbenet** [comment]
@mildred fantastic! :+1: :+1: :+1:  Let's get this merged in!

**Comment by jbenet** [comment]
Hm travis failed with this: https://travis-ci.org/jbenet/go-ipfs/jobs/48904581  (does this need adding stuff to `cmd/ipfs-gateway-fs` ? ) 

And sorry for the rebasing-- result of high-velocity dev. Do you want contributor access? it may make working on these larger pieces easier. I.e. i could push a commit to your branch that fixes that test.

**Comment by briantigerchow** [comment]
> Do you want contributor access?

:+1:

**Comment by mildred** [comment]
Thank you. I would be honoured to get contributor access. The work model would be to work on a branch on this repository, and everything else as usual, right? I'm not sure that I'll be as available as I am these days, though. And I have a few other projects as well.

Thank you for pointing me the travis log. I wasn't understanding how to find that info in travis until just now. I just updated the code to fix the build issue.

**Comment by mildred** [comment]
The fmt import was removed from cmd/ipfs/daemon.do in commit 73234479fa095f0a3fd01581ae1aee749d92f6bc and this broke the build. Re-updated the pull request (home it will pass this time)

**Comment by jbenet** [comment]
> Thank you. I would be honoured to get contributor access. 

Great, you've got it. 

> The work model would be to work on a branch on this repository, and everything else as usual, right? 

Correct. the only rules are:
- dont push straight to master (unless I ok it) 
- dont mess up other people's refs (unless they ok it) 
(basically, each branch has an "owner", and i own `master`)

Other notes:
- rebase often to stay on top of master
- pick long-ish branch names to avoid clashes
- if you rebase other people's branches, leave a comment in the associated PR (and be in contact with them)
- careful deleting refs (i.e. branches, tags) -- make sure nobody's using them / wants them

> I'm not sure that I'll be as available as I am these days, though. And I have a few other projects as well.

No worries

**Comment by jbenet** [comment]
I've a couple comments on the sharness tests, but I'll merge this now to avoid more divergence. rest can be a separate PR
